src/Rendering/GraphicsPipelineDesc.h

```diff
@@ -48,29 +97,176 @@ enum class CompareOp
+    Count
+};
+
 struct MultisampleState final
 {
-    uint32_t sampleCount = 1;
+    MultiSampleCount sampleCount = MultiSampleCount::One;
+    bool sampleShadingEnable     = false;
+    bool alphaToCoverageEnable   = false;
+    bool alphaToOneEnable        = false;
+    float minSampleShading       = 0.0f;
+    uint32_t sampleMask          = ~0u;
+};
+
+enum class StencilOp : uint8_t
+{
+    Keep,
+    Zero,
+    Replace,
+    IncrementAndClamp,
+    DecrementAndClamp,
+    Invert,
+    IncrementAndWrap,
+    DecrementAndWrap,
+    Count
+};
+
+struct StencilOpState final
+{
+    StencilOp failOp      = StencilOp::Replace;
+    StencilOp passOp      = StencilOp::Replace;
+    StencilOp depthFailOp = StencilOp::Replace;
+    CompareOp compareOp   = CompareOp::Never;
 };
 
 struct DepthStencilState final
 {
-    CompareOp depthTest = CompareOp::LessEqual;
+    // reverse depth buffer
+    bool depthTestEnable       = true;
+    bool depthWriteEnable      = true;
+    CompareOp depthTestOp      = CompareOp::GreaterEqual;
+    bool depthBoundsTestEnable = false;
+    bool stencilTestEnable     = false;
+    StencilOpState front       = StencilOpState();
+    StencilOpState back        = StencilOpState();
+    float minDepthBounds       = 0.0f;
+    float maxDepthBounds       = 1.0f;
 };
 
-struct RenderState final
+struct RasterizationState final
 {
+    bool depthClamp : 1            = false;
+    bool rasterizerDiscard : 1     = false;
+    bool frontCounterClockwise : 1 = false;
+    bool depthBiasEnable : 1       = false;
+    CullMode cullMode : 4          = CullMode::Back;
+    PolygonMode polygonMode : 8    = PolygonMode::Fill;
 };
 
-struct BlendState final
+static_assert(sizeof(RasterizationState) == 2, "RasterizationState is too big");
+
+struct ViewPort final
 {
+    float x;
+    float y;
+    float width;
+    float height;
+    float minDepth;
+    float maxDepth;
 };
 
-struct PipelineDesc final
+struct ViewPortState final
 {
-    const char* debugName = "Noname Pipeline";
+    int count           = 1;
+    ViewPort* viewPorts = nullptr;
+};
 
-    TopologyType topology = TopologyType::TriangleList;
+struct ScissorState final
+{
+    int count      = 1;
+    Rect* scissors = nullptr;
+};
+
+enum class BlendFactor : uint8_t
+{
+    Zero,
+    One,
+    SrcColor,
+    OneMinusSrcColor,
+    DstColor,
+    OneMinusDstColor,
+    SrcAlpha,
+    OneMinusSrcAlpha,
+    DstAlpha,
+    OneMinusDstAlpha,
+    ConstantColor,
+    OneMinusConstantColor,
+    ConstantAlpha,
+    OneMinusConstantAlpha,
+    SrcAlphaSaturate,
+    Src1Color,
+    OneMinusSrc1Color,
+    Src1Alpha,
+    OneMinusSrc1Alpha,
+    Count
+};
+
+// Symplified version of BlendOp
+enum class BlendOp : uint8_t
+{
+    Add             = 0,
+    Subtract        = 1,
+    ReverseSubtract = 2,
+    Min             = 3,
+    Max             = 4
+};
+
+enum class ColorComponent : uint8_t
+{
+    R = 1 << 0,
+    G = 1 << 1,
+    B = 1 << 2,
+    A = 1 << 3,
+    Count
```

Comment:
`Count` will be `1 << 3 + 1`. Is it intended?

Reply:
Remove Count now